src/Npgsql/NpgsqlConnector.cs

```diff
@@ -2002,6 +2011,11 @@ UserAction DoStartUserAction()
 
                 StartCancellableOperation(cancellationToken, attemptPgCancellation);
 
+                // We reset the UserTimeout for every user action, so it wouldn't leak from the previous query or action
+                // For example, we might have successfully cancelled the previous query (so the connection is not broken)
+                // But the next time, we call the Prepare, which doesn't set it's own timeout
+                UserTimeout = (command?.CommandTimeout ?? Settings.CommandTimeout) * 1000;
```

Comment:
I don't like all these conditions in the hot path :)

My idea was for startUserAction to receive the timeout as an (optional?) parameter, and then everyone who calls it supplies it. This way, COPY would also pass the timeout when opening the user action, and not directly on the buffer as now (I don't think we need to support modifying the timeout in the middle of a COPY operation). Wait is tricky because the timeout may change during the operation (for keepalive).

It's also not clear to me whether Settings.CommandTimeout should be a default for things which aren't commands...

Reply:
But we start user action for the COPY when it starts. Unless we add a param for it there...

Reply:
That's true. In that case, assuming we want to keep the current COPY timeout mechanism, I don't really see the point of moving the UserTimeout into StartUserAction - nobody actually uses it aside from the command, right?

Reply:
Every time we call connector's ReadMessage, we set the read buffer's timeout from the UserTimeout.

Reply:
Per offline discussion, we leave it as is. Made #3341 to track and fix later.